### DISCUSSION TITLE
UX for NULL values

### DISCUSSION THREAD

**Discussion by ghislaineguerin** [mathesar-foundation/mathesar#832] [Maintainer discussions]
### Problem
We need a design solution for handling NULL values in a way that helps users understand the concept of NULL as the 'absence of value' or 'value placeholder' rather than a 'blank or empty value.' This concept will be necessary when the user moves to more advanced features such as aggregation, formula, or expression-based columns. It will also matter for non-text data types, where a blank value will be considered invalid. For example, an email can be NULL but cannot be an empty string. 

### Proposed Solution
1. NULL values are represented by a cell containing the word NULL. A value can be deleted by pressing the 'Delete' key while focused on a cell. 
![image](https://user-images.githubusercontent.com/845767/142416645-e5074b7d-3d8a-4775-80b4-5c1785ce1b7e.png)

2. Empty String Values are automatically converted to NULL unless a substitution is determined
Data types that aren't text will most likely prevent users from entering a blank value as it would fail validation. We can, by default, set all empty values to NULL or have the user set a substitution for null, which can be 0 or an empty string, for example. 

3. Warnings are used to help users understand why they cannot insert NULL values in some cases
When constraints are applied to a column, a user might not be able to clear a value. We need to make sure they understand why the system is preventing the insert of NULL.

![image](https://user-images.githubusercontent.com/845767/142428510-6dd97579-7db4-45fc-80f6-bf72c5ca723e.png)

4. More descriptive value placeholders are used when a NULL cell is on focus
This is particularly important for types such as dates that require specific entry format or special inputs such as checkboxes for boolean type values

![image](https://user-images.githubusercontent.com/845767/142428908-445e7319-39d6-427a-8e47-599f0160cbaf.png)

## Questions
- Are there cases where a user might want to insert a blank value rather than NULL?
- Can a NULL value be copied and pasted into another cell turning it to NULL as well?
- In the case of one-to-one relationships created via the 'Link Table' dialog, a column with a UNIQUE constraint would be created, but contain all NULL values. Since they aren't unique, would that produce an error?
- Do we need to prototype any special NULL scenarios for specific data types?

**Comment by seancolsen** [comment]
> Can a NULL value be copied and pasted into another cell turning it to NULL as well?

Yes


> Are there cases where a user might want to insert a blank value rather than NULL?

I can't think of any reason why I'd design a schema that assigns meaning to an empty string separate from NULL. But my guess is that some people will be using Mathesar with schemata that do behave that way for whatever reason. They might have a separate UI that will break if the value is NULL. So I think we should still allow the user to enter an empty string. However in some cases it would be really nice to have Mathesar convert empty strings to NULL. So maybe the behavior could be configurable? I don't know how granular the configuration would need to be though.


> A value can be deleted by pressing the 'Delete' key while focused on a cell.

I like this feature. But I wonder if we also need to implement another mechanism for setting a value to NULL that is more discoverable and also usable on a touch device. Perhaps we have a context-menu (right click, or click-hold on touch devices) which gives the option "Set NULL".

> More descriptive value placeholders are used when a NULL cell is on focus

This seems a bit confusing. I like the idea of input masking, but only when the user is actively editing the cell.

**Reply by ghislaineguerin** [reply]
@seancolsen You make a good point about touch devices. I haven't been contemplating interactions for that. I think the contextual menu makes sense. 
As for the input, the idea was to show it while they are actively editing, maybe it wasn't clear from the description I've added.

**Reply by ghislaineguerin** [reply]
@seancolsen what do you think about having a 'Clear Value' option that simply clears the contents of a string type cell and it's the default behavior for the 'delete' key and another option 'Set as NULL' that would be accessible via the context menu as represented below. 
![image](https://user-images.githubusercontent.com/845767/143078134-375a7ac8-00cc-4754-bc4e-4592946134fa.png)

**Reply by seancolsen** [reply]
That looks great

Additional thoughts:

- "Clear Value" should only be available on some types. Likewise, "Set to NULL" should only be available when the column is nullable.
- We'll also need to consider that the user may have multiple cells selected across different columns when they activate the context menu. So each of those options should only be available if they option is available on all columns in the selection.
- Your mockup does make me wonder what will happen if I choose "Cut". Will it the cell end up as NULL or empty string? I think we can take up that question when we get into the clipboard feature.

**Reply by pavish** [reply]
Having a context menu option for 'Set to NULL' is a great idea. We can show that option when the column is nullable.

However, I don't think we need 'Clear Value' on the context menu. Users can edit the cell and remove the value if they wish to clear. Clear only applies for textual columns where it would be cleared to an empty string, for all other column types it's either a valid value or null.

**Comment by kgodey** [comment]
> In the case of one-to-one relationships created via the 'Link Table' dialog, a column with a UNIQUE constraint would be created, but contain all NULL values. Since they aren't unique, would that produce an error?

Columns with unique constraints can have multiple records with NULL values. NULL is not treated as a value; it's an absence of a value so it doesn't make sense to include it in the unique comparison.

**Reply by ghislaineguerin** [reply]
Thanks @kgodey, this answers my question.

**Comment by mathemancer** [comment]
> Are there cases where a user might want to insert a blank value rather than NULL?

Yes.  Not allowing the empty string for text columns is equivalent to not allowing zero for numeric columns when considering any operations involving string concatenation, slicing, equality testing, etc.  

Example:

User wants a generated column that is true if the top-level-domain for `sender_email` and `receiver_email` are equal.  They have `sender_email_tld` and `receiver_email_tld` columns, and the generated column is just generated by
  ```sql
  SELECT sender_email_tld = receiver_email_tld
  ```
  They may want to consider addresses `@localhost` to have equal TLDs.  If we set empty strings to `NULL`, they'll instead have `NULL` for those entries in that column, since `SELECT NULL = NULL` returns `NULL`, not `True`.

I even think it would be unwise to encourage users to automatically set empty strings to `NULL`, since this could create a number of confusing results if they don't understand the implications for behavior under various operators.  Again, it's equivalent to not having zero in numeric columns.  Over time, we'll eventually be helping the user do more and more data processing and manipulation in Mathesar, and removing the identity element from a type would make that quite difficult due to these behaviors under operators. 

> Can a NULL value be copied and pasted into another cell turning it to NULL as well?

Yes, I think makes sense.  It's not "technically correct", but I don't see any potential for creating mistakes or confusion (at the moment).

> In the case of one-to-one relationships created via the 'Link Table' dialog, a column with a UNIQUE constraint would be created, but contain all NULL values. Since they aren't unique, would that produce an error?

The usual way to do one-to-one relationships would be to make the primary key of a given table also a foreign key.  In that case, it will always be non-null and unique.  I suggest sticking with that unless we have a very good reason to move away from it.  If we really want to move away from it, we need to spend some effort understanding the implications.

**Reply by ghislaineguerin** [reply]
@mathemancer, you make a good point about empty strings. If we don't automatically transform empty to NULL, then there's almost no reason for a user to set a column to NULL intentionally, is there? Unless they want to concatenate values from those columns. Are there any other reasons that you can think of where empty vs. NULL would affect the experience of users who aren't aware of the differences?

**Reply by mathemancer** [reply]
`NULL` values would be useful in the case of a column with a unique constraint (since you could defer adding values; `NULL` doesn't break uniqueness).  Also, giving the user the option might be useful in my example above.  Perhaps they _do_ want a nullable boolean for whether TLDs are equal; i.e., no TLD means no answer for that, i.e., set the `is_equal` column to `NULL`.  There's also some use when it comes to multicolumn foreign keys, and when counting values (both counting distinct and counting with duplicates).

One thing that is opened if we want both `NULL` and empty is how to let them specify each.  I guess if they hit "delete" on the whole cell, it goes to `NULL`, but then if they want an empty string they just backspace the characters out of the cell in "edit mode"?

`NULL` values also behave differently from empty strings (or zero) when sorting.  They can be chosen to be either the maximum or minimum value (this is a DB config, can be set for a column, and can be chosen ad-hoc when sorting as well).

**Reply by ghislaineguerin** [reply]
thanks, @mathemancer this is helpful. I guess my question was more about how a user who has none of those problems will first come into contact with the concept of NULL. Seems most of your examples would apply to a user that has already moved into setting up table relationships. I assume the first time they'd interact with NULLS is when:

- They add a new column to an existing table
- They add a new record and leave some fields without values
- They set a different data type to a column with existing values and they are cleared
- They delete values for a data type that does not support empty strings (dates, email, boolean)

It seems impractical to add an empty string by backspacing characters. Could it be that for text types, entering edit mode automatically makes it not NULL? They can always hit delete to remove the empty string. The problem with this is that it would lead to errors in other types that require validation, for example, '' as an invalid email or URL. Since the user didn't type anything, they'd be confused as to why we are validating the value of the field.

**Reply by mathemancer** [reply]
Some users may want to attach semantic meaning to `NULL` in the scenarios you mentioned, but I wouldn't really encourage it.  For example, they may interpret empty string as "the student has no middle name", whereas `NULL` in that field would be "the information of whether the student has a middle name and what it might be is missing".  Generally, however, I think `NULL` vs. empty string is most important when processing and working with data, moreso than inputting and reading data by itself.

**Comment by pavish** [comment]
> A value can be deleted by pressing the 'Delete' key while focused on a cell

This is a good idea. I think we can convert the value to NULL when user presses 'Delete' key on a selected cell. However, if the user double clicks into edit mode, and deletes everything using Backspace or Delete, we would end up with an empty string which we can retain. This only applies for text based fields, everything else would default to NULL.

> Can a NULL value be copied and pasted into another cell turning it to NULL as well?

This is a bit tricky. Consider the user copies NULL value, we should allow them to paste it on other spreadsheet applications/text editors as well, where it will be pasted as a CSV. So, would we have a NULL in the CSV? Or would we have empty csv field? Having NULL would be confusing since it could be interpreted as a string with value 'NULL'.

**Reply by seancolsen** [reply]
Copying from Mathesar and pasting into another application is a really interesting problem! I don't think NULL values will be the only challenges therein. For example, if the user copies a date, what's the expected format that we use when pasting into another application? I'd suggest we address these challenges in detail once we're building the copy-paste functionality.

**Reply by pavish** [reply]
One of our usecases is being able to copy from a spreadsheet and paste into Mathesar. So, we should support it both ways. Ideally, the content of the copied cells would be in a CSV format.

**Reply by seancolsen** [reply]
> Ideally, the content of the copied cells would be in a CSV format

I've never built any custom clipboard functionality into a web app and I don't know much about how clipboards work in general, but I know that in some applications I can copy content which pastes differently depending on whether I'm pasting into that application or into another application. I figure there must be a way to achieve the same behavior with Mathesar. 

Here's an example of copy-paste within a Google Sheet

![copy-note](https://user-images.githubusercontent.com/42411/142635497-7bae3ef1-a6e2-4fc0-a164-c7a1fef9b949.gif)

Pasting within the sheet pastes the note. Pasting to a plain text document does not paste the note. I think we'd do something similar for NULL values.

**Comment by kgodey** [comment]
I don't have much feedback to add, I think everyone else has it covered.
